### PR TITLE
Add SHOW PRINCIPALS and SHOW ROLE GRANTS FOR principal

### PR DESCRIPTION
**Pull request by lhofhansl**
Apologies for the inundation with changes...

This adds the ability to list all principals for given role, as well as the ability to list the roles granted to a principal.

This WIP, mostly get a look at the potential syntax.

Some discussion on: #3167

### CODE REVIEW

**Comment by kokosing** [comment]
@martint Please review the syntax, I can take of the rest.

**Comment by lhofhansl** [comment]
Finished SHOW ROLE GRANTS (FOR | TO) principal with security, etc. Tests still missing.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/security/SqlStandardAccessControl.java [hunk 1]
```diff
@@ -402,8 +403,26 @@ public void checkCanShowCurrentRoles(ConnectorSecurityContext context, String ca
...
+        }
+
+        // Follow what Hive is doing:
+        //   users can find out about themselves...
+        ConnectorIdentity identity = context.getIdentity();
+        if (principal.getType() == USER && identity.getUser().equals(principal.getName())) {
+            return;
+        }
+
+        //   ... and about roles granted to them (but not necessary enabled)
```
Comment (lhofhansl) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/security/SqlStandardAccessControl.java]: "necessarily" I'll fix the type in next push

**Comment by lhofhansl** [comment]
@martint , any thoughts on the syntax?

**Comment by lhofhansl** [comment]
I have cycles to go on with this - but only if it's the right direction. @kokosing @martint

**Comment by lhofhansl** [comment]
Any interest?

I'd rather not create an internal fork for this, but we need functionality like this.

I'm not attached to the syntax, but to the functionality, the ability to query the access metadata, and to do so without needing to go back to Hive. Could do via the information_schema if that's better.

**Comment by lhofhansl** [comment]
With #3099 merged, this is the last one on this set of related changes. With all these in place we have fairly comprehensive support for SQL authorization directly in Presto (without the need to defer these tasks to Hive).

@kokosing , @martint Could I get you to have a look at the direction and syntax.
Much of this is not implemented, yet. I just want to discuss the syntax, etc.

**Comment by kokosing** [comment]
I would prefer to have dedicated tables in `information_schema` that are much useful (you can join, or filter). Then `SHOW ...` could be a syntactic sugar on top of it (if it is still needed).

**Comment by lhofhansl** [comment]
Agreed. I'll figure out how to do that.

**Comment by lhofhansl** [comment]
How about new `role_grants` table in the `information_schema`?
The table's schema would simple be `principal, role` and is only accessible by an admin.

So through this table one can thus find roles by principal or principals by role (i.e. what `SHOW PRINCIPALS group` or `SHOW ROLE GRANTS FOR principal` would do).

Unfortunately there is no standard to follow here - that I know off.

**Comment by kokosing** [comment]
Notice that `information_schema` provides tables like `applicable_roles` and `enabled_roles`. The content of these tables is dependant on session (current user). Former represents `SHOW ROLES` and latter `SHOW CURRENT ROLES`. Please see: `io.prestosql.sql.rewrite.ShowQueriesRewrite.Visitor#visitShowRoles`

**Comment by kokosing** [comment]
I guess my previous comment was not much helpful here.

> exceedingly expensive.

We could implement predicate pushdown so we could only request information that is needed.

**Comment by lhofhansl** [comment]
> We could implement predicate pushdown so we could only request information that is needed.

Lemme look into that, that'd be cool.
Do any of the `information_schema` tables do that already so that I can look at an example?

If no predicate is provided do we let it through anyway and accept the load on the metastore?

**Comment by kokosing** [comment]
Yes. There is generic predicate pushdown support for catalog, schema and table name columns that works for every connector. See `InformationSchemaPageSource`. You would need add support for other columns too. It shouldn’t be very hard.

Yes. We are ok with the fact that full scan is expensive. It is quite typical with `information_schema` tables.

**Comment by lhofhansl** [comment]
Going to close this one in favour of #3535

**Pull request closed without merging**